### PR TITLE
Unix Domain Sockets HttpClientBuilder fixes

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
UDS is supported by ServicetTalk by passing in a DomainSocketAddress
address. However the HttpClientBuilder had some restrictions/assumptions
about the address type which prevented its use.

Modifications:
- Relax assumptions/restrictions in the HttpClientBuilder to allow for
DomainSocketAddress
- Add unit test to demonstrate UDS
- Add example for UDS

Result:
UDS client is now working and tested.

### CODE REVIEW

File: servicetalk-examples/http/uds/build.gradle [hunk 1]
```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright © 2019 Apple Inc. and the ServiceTalk project authors
```
Comment (idelpivnitskiy) [inline servicetalk-examples/http/uds/build.gradle:2]: Please update the copyright year in all files to 2020.

File: servicetalk-examples/http/uds/src/main/java/io/servicetalk/examples/http/uds/blocking/BlockingUdsServer.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+public final class BlockingUdsServer {
+    public static void main(String[] args) throws Exception {
+        DomainSocketAddress udsAddress = udsAddress();
+        try {
+            HttpServers.forAddress(udsAddress)
+                    .listenBlockingAndAwait((ctx, request, responseFactory) ->
+                            responseFactory.ok().payloadBody("Hello World!", textSerializer()))
+                    .awaitShutdown();
+        } finally {
+            new File(udsAddress.getPath()).delete(); // After the server is done, clean up the file.
```
Comment (idelpivnitskiy) [inline servicetalk-examples/http/uds/src/main/java/io/servicetalk/examples/http/uds/blocking/BlockingUdsServer.java:38]: May be in a separate PR: can you please remove `get` prefix from `DomainSocketAddress#getPath()` method for consistency with other API?

File: servicetalk-examples/http/uds/src/main/java/io/servicetalk/examples/http/uds/blocking/UdsUtils.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+     * Create a {@link DomainSocketAddress} that is backed by a fixed file such that different JVMs
+     * can use it to communicate via UDS.
+     * @return a {@link DomainSocketAddress} that is backed by a fixed file such that different JVMs
+     * can use it to communicate via UDS.
+     */
+    public static DomainSocketAddress udsAddress() {
+        final String tempDirProp = "java.io.tmpdir";
+        final String tempDir = System.getProperty(tempDirProp);
+        if (tempDir == null) {
+            throw new IllegalStateException("unable to find " + tempDirProp + " in System properties");
```
Comment (idelpivnitskiy) [inline servicetalk-examples/http/uds/src/main/java/io/servicetalk/examples/http/uds/blocking/UdsUtils.java:34]: Consider using `File#createTempFile` instead of requiring a system property to simplify the example.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java [hunk 4]
```diff
@@ -495,6 +495,7 @@ HttpExecutionStrategyInfluencer buildStrategyInfluencerForClient(HttpExecutionSt
         return influencerChainBuilder.buildForClient(strategy);
     }
 
+    @Nullable
```
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java:498]: Why the `@Nullable` required now?

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java [hunk 5]
```diff
@@ -179,9 +179,9 @@ private DefaultSingleAddressHttpClientBuilder(@Nullable final U address,
                 globalDnsServiceDiscoverer());
     }
 
-    static <U> DefaultSingleAddressHttpClientBuilder<U, InetSocketAddress> forResolvedAddress(
-            final U u, final InetSocketAddress address) {
-        ServiceDiscoverer<U, InetSocketAddress, ServiceDiscovererEvent<InetSocketAddress>> sd =
+    static <U, R> DefaultSingleAddressHttpClientBuilder<U, R> forResolvedAddress(
```
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java:182]: Should `R` extend `SocketAddress`?

**Review by idelpivnitskiy — COMMENTED** [review]
Good catch!

File: servicetalk-examples/http/uds/src/main/java/io/servicetalk/examples/http/uds/blocking/UdsUtils.java (hunk 3 above)
Comment (tkountis) [reply servicetalk-examples/http/uds/src/main/java/io/servicetalk/examples/http/uds/blocking/UdsUtils.java:34]: I believe a fixed name is needed, not a random one, the `File#createTempFile` wouldn't work in that case

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java (hunk 5 above)
Comment (Scottmitch) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java:182]: done. somewhat on the fence if we need to take a dependency on the marker base class in our APIs, but we should be able to relax the generic constraints later if necessary.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java (hunk 4 above)
Comment (Scottmitch) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java:498]: good call, not needed. was returning `null` at one point during development.

File: servicetalk-examples/http/uds/src/main/java/io/servicetalk/examples/http/uds/blocking/UdsUtils.java (hunk 3 above)
Comment (Scottmitch) [reply servicetalk-examples/http/uds/src/main/java/io/servicetalk/examples/http/uds/blocking/UdsUtils.java:33]: +1 to what @tkountis said. Random file name is not suitable for this use case. The system property is set by the JVM [1] and this code is just a bit defensive.

[1] https://docs.oracle.com/javase/8/docs/api/java/lang/System.html

File: servicetalk-examples/http/uds/src/main/java/io/servicetalk/examples/http/uds/blocking/BlockingUdsServer.java (hunk 2 above)
Comment (Scottmitch) [reply servicetalk-examples/http/uds/src/main/java/io/servicetalk/examples/http/uds/blocking/BlockingUdsServer.java]: Separate PR, I'm not sure this is worth breaking APIs over.

Comment (idelpivnitskiy) [reply servicetalk-examples/http/uds/src/main/java/io/servicetalk/examples/http/uds/blocking/BlockingUdsServer.java]: Sounds good, looks like `DomainSocketAddress` was not useful before this PR, should be ok to change API.

**Review by idelpivnitskiy — APPROVED** [review]
🎖️
